src/python/pants/backend/python/tasks2/build_local_python_distributions.py

```diff
@@ -0,0 +1,85 @@
+from pants.base.build_environment import get_buildroot
+from pants.base.exceptions import TaskError
+from pants.base.fingerprint_strategy import DefaultFingerprintStrategy
+from pants.task.task import Task
+from pants.util.dirutil import safe_mkdir
+
+
+class BuildLocalPythonDistributions(Task):
+  """Create python distributions (.whl) from python_dist targets."""
+
+  options_scope = 'python-create-distributions'
+  PYTHON_DISTS = 'user_defined_python_dists'
+
+  @classmethod
+  def product_types(cls):
+    return [cls.PYTHON_DISTS]
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    round_manager.require_data(PythonInterpreter)
+
+  @staticmethod
+  def is_distribution(target):
+    return isinstance(target, PythonDistribution)
+
+  @property
+  def cache_target_dirs(self):
+    return True
+
+  def execute(self):
+    dist_targets = self.context.targets(self.is_distribution)
+    built_dists = set()
+    
+    if dist_targets:
+      with self.invalidated(dist_targets,
+                            fingerprint_strategy=DefaultFingerprintStrategy(),
+                            invalidate_dependents=True) as invalidation_check:
+        for vt in invalidation_check.all_vts:
+          if vt.valid:
+            built_dists.add(self._get_whl_from_dir(os.path.join(vt.results_dir, 'dist')))
+          else:
+            built_dists.add(self._create_dist(vt.target, vt.results_dir))
+
+    self.context.products.register_data(self.PYTHON_DISTS, built_dists)
+
+  def _create_dist(self, dist_tgt, dist_target_dir):
+    """Create a .whl file for the specified python_distribution target."""
+    interpreter = self.context.products.get_data(PythonInterpreter)
+
+    # Copy sources and setup.py over to vt results directory for packaging.
+    # NB: The directory structure of the destination directory needs to match 1:1
+    # with the directory structure that setup.py expects.
+    for src_relative_to_target_base in dist_tgt.sources_relative_to_target_base():
+      src_rel_to_results_dir = os.path.join(dist_target_dir, src_relative_to_target_base)
+      safe_mkdir(os.path.dirname(src_rel_to_results_dir))
+      abs_src_path = os.path.join(get_buildroot(),
+                                  dist_tgt.address.spec_path,
+                                  src_relative_to_target_base)
+      shutil.copyfile(abs_src_path, src_rel_to_results_dir)
+    # Build a whl using SetupPyRunner and return its absolute path.
+    setup_runner = SetupPyRunner(dist_target_dir, 'bdist_wheel', interpreter=interpreter)
+    setup_runner.run()
+    return self._get_whl_from_dir(os.path.join(dist_target_dir, 'dist'))
+
+  def _get_whl_from_dir(self, install_dir):
+    """Return the absolute path of the whl in a setup.py install directory."""
+    dists = glob.glob(os.path.join(install_dir, '*.whl'))
+    if len(dists) == 0:
+      raise TaskError('No distributions were produced by python_create_distribution task.')
```

Comment:
producing >1 wheels should probably be an error condition, or at least a warning - vs the silent masking happening here now.

Reply:
✅